src/main/java/com/github/firmwehr/gentle/source/Source.java

```diff
@@ -1,40 +1,88 @@
 package com.github.firmwehr.gentle.source;
 
-import java.util.List;
+import com.github.firmwehr.gentle.util.Pair;
+
 import java.util.Objects;
-import java.util.stream.Collectors;
 
 public class Source {
 
 	private final String content;
-	private final List<String> lines;
 
 	public Source(String content) throws SourceException {
 		if (!content.codePoints().allMatch(c -> c <= 127)) {
 			throw new SourceException("input contains non-ASCII characters");
 		}
 
 		this.content = content;
-		this.lines = content.lines().collect(Collectors.toList());
 	}
 
 	public String getContent() {
 		return content;
 	}
 
-	public String formatErrorAtPosition(SourcePosition position, String message, String description) {
-		String line;
-		if (position.line() - 1 < lines.size()) {
-			line = lines.get(position.line() - 1);
-		} else {
-			line = "";
+	public Pair<SourcePosition, String> positionAndLineFromOffset(int offset) {
```

Comment:
I feel like this logic can be simplified. As it is now, it feels difficult to follow.